src/server/qgsconfigcache.h

```diff
@@ -30,20 +31,59 @@
 #include "qgsproject.h"
 #include "qgsserversettings.h"
 
+#ifndef SIP_RUN
+
+
+#endif
+
+class QgsConfigCache;
+
+/**
+ * \ingroup server
+ * \brief Abstract cache strategy for server configuration
+ * \since QGIS 3.24
+ */
+class SERVER_EXPORT QgsAbstractCacheStrategy  : public QObject
+{
+    friend class QgsConfigCache;
+
+    Q_OBJECT
+  public:
+    virtual QString ident() const = 0;
+
+  protected:
+
+    /**
+     * Called when an entry is removed from cache.
+     * \param path The path of the project
+     */
+    virtual void entryRemoved( const QString &path ) = 0;
+
+    //! Called when an entry is inserted
+    virtual void entryInserted( const QString &path ) = 0;
+
+    //! Attache cache to this strategy
+    virtual void attach( QgsConfigCache *cache ) = 0;
+};
+
+
 /**
  * \ingroup server
  * \brief Cache for server configuration.
  * \since QGIS 2.8
  */
 class SERVER_EXPORT QgsConfigCache : public QObject
 {
+    friend class QgsFileSystemCacheStrategy;
+    friend class QgsPeriodicCacheStrategy;
+
     Q_OBJECT
   public:
 
     /**
      * Returns the current instance.
      */
-    static QgsConfigCache *instance();
+    static QgsConfigCache *instance( QgsServerSettings *settings = nullptr );
```

Comment:
I believe this is mixing concerns of initialisation and acquiring an instance. Better separate these two tasks.

Reply:
An instance should not be created with no strategy, because  `QgsConfigCache::instance()` is public api and users expect
to get the default strategy.

In the context  of the server inititalisation this would mean:
1- Create the instance with the default strategy (allocate an object)
2- Immediately after,  change the strategy according to the settings (destroy the previous object and reallocate another strategy
 
That's sound not very efficient - but I can live with that.

Another option would be to create another function `createInstanceWithOptions`  called from within the server `init()` only.

Reply:
The second thing is what I was thinking of -- you could even add a `Q_ASSERT( settings )` in `instance()` to complain if someone tries to access instance before init.

Reply:
>  you could even add a Q_ASSERT( settings ) in instance() to complain if someone tries to access instance before init.

This may be considered as an API break or I'm wrong ?

Reply:
Q_ASSERT is a debug build safety belt only, replace it with other logging mechanism if you prefer. Also I don't see that happen in the wild, this will always be initialized internally.
Crashing / excepting on a misuse of API is IMHO always preferable over UB.

Reply:
> Also I don't see that happen in the wild

As long as QgsConfigCache is not used outside the context of the server `init()`. I think that @elpaso use it in some  scripts.

Reply:
> Q_ASSERT is a debug build safety belt only, replace it with other logging mechanism

Ok, so that won't prevent instance creation, then we have to return an instance with the default strategy.

Reply:
If we say "It is required to call init() before the first call to instance(), not doing so is undefined behavior" we do not need to create the default strategy (and we help developers working in developer mode to find this by Q_ASSERT'ing).

Reply:
>  not doing so is undefined behavior

Not really , because it always define a default strategy. There is a also log message to warn that init() has not been called and that the default behavior is used.